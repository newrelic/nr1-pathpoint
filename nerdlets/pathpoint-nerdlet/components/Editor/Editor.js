// React
import React from 'react';

// Librarys
import PropTypes from 'prop-types';

// Assets
import editorFilters from '../../config/nrql-keywords.json';

const Editor = ({ name, style, value, isReadOnly, onChange, onPressEnter }) => {
  const refBackdrop = React.useRef();
  const refCustomArea = React.useRef();
  const refTextarea = React.useRef();

  // Create object with keyword (property) and color (value)
  const keywords = React.useMemo(() => {
    return editorFilters.keys.reduce((acc, el) => {
      const currentKeys = {};

      // Iterate each filter keyword
      el.keywords.forEach(keyword => {
        // Asign to current keys, the keyword and the respective color
        Object.assign(currentKeys, {
          [keyword]: el.color
        });
      });

      return { ...acc, ...currentKeys };
    }, {});
  }, []);

  /*
    Parse special keywords as '*', add to special keyword the '/' token before the keyword,
    For example { '*': '#000000' } =====> { '\*': '#000000' }
  */
  const parsedKeywords = React.useMemo(() => {
    const keys = Object.keys(keywords); // Get properties from keywords
    const backupKeywords = { ...keywords }; // Create keywords backup

    // Iterate each special character
    editorFilters.specialCharacters.forEach(specialCharacter => {
      // Match any special character prop with any keyword in filters prop
      const existKeyInSpecialCharacters = keys.some(
        key => key === specialCharacter
      );

      // If its match, add to keyword, the '\' token
      if (existKeyInSpecialCharacters) {
        delete backupKeywords[specialCharacter]; // Delete property
        const parsedKey = `\\${specialCharacter}`; // Add '\' token
        backupKeywords[parsedKey] = keywords[specialCharacter]; // Set new property parsed
      }
    });

    return backupKeywords;
  }, []);

  const applyColors = React.useCallback(text => {
    // Define regex for match the NRQL keywords in the text
    const regx = new RegExp(Object.keys(parsedKeywords).join('|'), 'gi');
    // Replace the NRQL keyword with html element
    return text.replace(regx, function(m) {
      const keywordColor = keywords[m.toLowerCase()]; // Get color for set to keyword
      return `<span style="color:${keywordColor}">${m}</span>`; // Return html element
    });
  }, []);

  const updateEditorText = React.useCallback(text => {
    refCustomArea.current.innerHTML = applyColors(text);
  }, []);

  // Event 'onChange' in textarea
  const handleOnInput = React.useCallback(e => {
    updateEditorText(e.target.value);
  }, []);

  // Event 'scroll' in textarea
  const handleOnScroll = React.useCallback(() => {
    refBackdrop.current.scrollTop = refTextarea.current.scrollTop;
  }, [refBackdrop, refTextarea]);

  // Event 'keyDown' for get current key pressed
  const handleOnKeyDown = React.useCallback(e => {
    if (e.keyCode !== 13) return false;

    e.preventDefault();

    const isFunc = typeof onPressEnter === 'function';

    if (!isFunc) return false;

    onPressEnter();
  }, []);

  React.useEffect(() => {
    let isMounted = true;

    if (isMounted) {
      updateEditorText(value);
    }

    return () => {
      isMounted = false;
    };
  }, [value]);

  return (
    <div className="nrql-editor" style={style}>
      <textarea
        ref={refTextarea}
        className="entry"
        spellCheck="false"
        autoCorrect="false"
        name={name}
        value={value}
        onChange={onChange}
        readOnly={isReadOnly}
        onInput={handleOnInput}
        onScroll={handleOnScroll}
        onKeyDown={handleOnKeyDown}
        style={
          isReadOnly
            ? { overflow: 'hidden', display: 'none' }
            : { whiteSpace: 'break-spaces' }
        }
      />

      <div ref={refBackdrop} className="backdrop">
        <div
          ref={refCustomArea}
          className="custom-area"
          style={!isReadOnly ? { whiteSpace: 'break-spaces' } : null}
        />
      </div>
    </div>
  );
};

export default React.memo(Editor);

Editor.propTypes = {
  name: PropTypes.string,
  style: PropTypes.object,
  value: PropTypes.string,
  isReadOnly: PropTypes.bool,
  onChange: PropTypes.func,
  onPressEnter: PropTypes.func
};
