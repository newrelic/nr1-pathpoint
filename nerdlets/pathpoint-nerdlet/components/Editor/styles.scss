.nrql-editor {
    position: relative;
    background-color: #ffffff;

    & > .backdrop {
        z-index: 1;
        overflow: auto;
        position: absolute;

        & .custom-area {
            width: inherit;
            height: inherit;
        }
    }

    & > .entry {
        z-index: 2;
        outline: 0;
        border-radius: 0 !important;
        resize: none !important;
        position: absolute !important;
        color: transparent !important;
        background-color: transparent !important;
        caret-color: #555555;
        min-width: 100% !important;
        max-width: 100%;
        border-width: 0 !important;
    }

    & > .backdrop, > .backdrop .custom-area, > .entry {
        width: 100%;
        height: 100%;
    }
    
    & > .backdrop .custom-area, > .entry {
        font-size: 14px !important;
        line-height: 22px !important;
        padding: 4px 7px !important;
        letter-spacing: .4px !important;
        word-break: break-word !important;
        font-family: 'Open Sans', sans-serif !important;
    }
}