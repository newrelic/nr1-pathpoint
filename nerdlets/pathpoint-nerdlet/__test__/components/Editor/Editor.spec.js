import React from 'react';
import { create } from 'react-test-renderer';
import Editor from '../../../components/Editor/Editor';

describe('<Editor />', () => {
  const name = 'Editor';
  const style = {
    width: '85px',
    heigth: '70px',
    backgroundColor: '#FFFFFF'
  };
  const value = 'value';
  const defaultValue = 'Previous value';
  const isReadOnly = true;
  const onChange = jest.fn();
  const onPressEnter = jest.fn();
  it('Render Editor correctly', () => {
    const wrapper = create(
      <Editor
        name={name}
        style={style}
        value={value}
        defaultValue={defaultValue}
        isReadOnly={isReadOnly}
        onChange={onChange}
        onPressEnter={onPressEnter}
      />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
