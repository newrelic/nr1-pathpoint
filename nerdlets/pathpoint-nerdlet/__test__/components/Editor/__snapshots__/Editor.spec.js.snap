// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Editor /> Render Editor correctly 1`] = `
<div
  className="nrql-editor"
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "heigth": "70px",
      "width": "85px",
    }
  }
>
  <textarea
    autoCorrect="false"
    className="entry"
    name="Editor"
    onChange={[MockFunction]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onScroll={[Function]}
    readOnly={true}
    spellCheck="false"
    style={
      Object {
        "display": "none",
        "overflow": "hidden",
      }
    }
    value="value"
  />
  <div
    className="backdrop"
  >
    <div
      className="custom-area"
      style={null}
    />
  </div>
</div>
`;
